### PR TITLE
chore: improve secretmanager on edge case handling

### PR DESCRIPTION
**Pull request by yuwenma**
### Change description

Fixes #

### Tests you have done

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/basicsecretmanagersecret/_generated_object_basicsecretmanagersecret.golden.yaml [hunk 1]
```diff
@@ -19,14 +20,13 @@ spec:
     userManaged:
       replicas:
       - location: us-central1
+  resourceID: secretmanagersecret-${uniqueId}
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/basicsecretmanagersecret/_generated_object_basicsecretmanagersecret.golden.yaml]: TF always add the `spec.resourceID` even if `state-into-spec: absent`

File: pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/basicsecretmanagersecret/_http.log [hunk 2]
```diff
@@ -24,14 +23,13 @@ X-Xss-Protection: 0
...
+POST https://secretmanager.googleapis.com/v1/projects/${projectId}/secrets?alt=json&secretId=secretmanagersecret-${uniqueId}
 Content-Type: application/json
-User-Agent: kcc/controller-manager
-x-goog-request-params: parent=projects%2F${projectId}
+User-Agent: Terraform/ (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google-beta/kcc/controller-manager
 
 {
   "labels": {
-    "label-one": "value-one",
+    "cnrm-test": "true",
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/basicsecretmanagersecret/_http.log]: (regression?) TF-based uses `metadata.labels` + `managed-by-cnrm`  - system labels  , the direct uses `spec.labels` + `managed-by-cnrm`.

File: pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/secretmanagersecretversion/_generated_object_secretmanagersecretversion.golden.yaml [hunk 3]
```diff
@@ -3,17 +3,20 @@ kind: SecretManagerSecretVersion
 metadata:
   annotations:
     cnrm.cloud.google.com/management-conflict-prevention-policy: none
+    cnrm.cloud.google.com/observed-secret-versions: (removed)
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/secretmanagersecretversion/_generated_object_secretmanagersecretversion.golden.yaml]: This is a legacy annotation. The TF-based approach builds a mechanism  around this annotation to determine if a sensitive data is changed.

File: pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/secretmanagersecretversion/_generated_object_secretmanagersecretversion.golden.yaml [hunk 4]
```diff
@@ -3,17 +3,20 @@ kind: SecretManagerSecretVersion
...
   - cnrm.cloud.google.com/deletion-defender
-  generation: 2
+  generation: 3
   labels:
     cnrm-test: "true"
   name: secretmanagersecretversion-${uniqueId}
   namespace: ${uniqueId}
 spec:
   enabled: false
+  resourceID: "1"
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/secretmanagersecretversion/_generated_object_secretmanagersecretversion.golden.yaml]: secret version is a service-genereated-ID resource, that TF_based always write back to `spec.resourceID`. in direct, we use `status.externalRef` instead.

File: pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/basicsecretmanagersecret/_generated_object_basicsecretmanagersecret.golden.yaml
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/secretmanager/v1beta1/basicsecretmanagersecret/_generated_object_basicsecretmanagersecret.golden.yaml:1]: Besides the basic test suite, I also added full test suite, where the TF-based controller cannot reconcile correctly because it assume a mutable field is immutable. 
Here's [the error](https://screenshot.googleplex.com/7W8MxZsxeE527sV.png), and here's [the tracking issue](https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/3051) with more details.

File: pkg/controller/direct/secretmanager/secretversion_controller.go [hunk 5]
```diff
@@ -164,6 +164,19 @@ func (a *SecretVersionAdapter) Create(ctx context.Context, createOp *directbase.
 	}
 	log.V(2).Info("successfully created SecretVersion", "name", a.id)
 
+	// The default status for newly created resource is "enabled".
```
Comment (justinsb) [inline pkg/controller/direct/secretmanager/secretversion_controller.go:167]: This is a bummer, but I verified that this is indeed the behaviour!

File: apis/secretmanager/v1beta1/secret_reference.go [hunk 6]
```diff
@@ -72,9 +72,16 @@ func (r *SecretRef) NormalizedExternal(ctx context.Context, reader client.Reader
...
-		return "", fmt.Errorf("reading status.externalRef: %w", err)
+	actualExternalRef, _, err1 := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err1 != nil {
+		err1 = fmt.Errorf("SecretManagerSecret `status.externalRef` not configured: %w", err1)
+		// Backward compatible to Terraform/DCL based resource, which does not have status.externalRef.
+		var err2 error
+		actualExternalRef, _, err2 = unstructured.NestedString(u.Object, "status", "name")
+		if err2 != nil {
+			err2 = fmt.Errorf("SecretManagerSecret `status.name` not configured: %w", err2)
+			return "", fmt.Errorf("%w\n%w", err1, err2)
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply apis/secretmanager/v1beta1/secret_reference.go:83]: Good call! Thanks!

File: pkg/controller/direct/secretmanager/secret_controller.go [hunk 7]
```diff
@@ -262,6 +239,10 @@ func (a *Adapter) Update(ctx context.Context, op *directbase.UpdateOperation) er
 	if mapCtx.Err() != nil {
 		return mapCtx.Err()
 	}
+	// Set externalRef. This field is empty after migration from the TF-based controller to the direct.
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply pkg/controller/direct/secretmanager/secret_controller.go:262]: Secret is not service generated ID resource (SecretVersion does). So it can switch back to use the TF-based controller. The real problem on using the TF-based controller is that the Secret `spec.rotation.rotationPeriod` is [broken](https://screenshot.googleplex.com/7W8MxZsxeE527sV.png). If user configures that fields in Direct and switched back to TF-based, it no longer works.
